xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/Logic.js

```diff
@@ -126,6 +164,277 @@ define(["jquery"], function ($) {
+    if (level !== -1) {
+      this.data.query.sort.splice(level, 1, sortObject);
+    }
+    if (currentLevel !== -1 && currentLevel !== level) {
+      this.data.query.sort.splice(currentLevel, 1);
+    }
+    // dispatch events
+    this._triggerEvent("sort", {
+      property: property,
+      level: level,
+      descending: descending,
+    });
+
+    // CALL FUNCTION TO FETCH NEW DATA HERE
+  };
+
+
+  /**
+   * Add new sort entry, shorthand of Logic.prototype.sort
+   * If the property is already sorting, does nothing
+   * @param {String} property The property to add to the sort
+   * @param {String} descending Specify whether the sort should be descending or not.
+   *   Undefined means toggle current direction
+   */
+  Logic.prototype.addSort = function (property, descending) {
+    var currentLevel = -1;
+    this.data.query.sort.some(function (sortObject, i) {
+      if (sortObject.property === property) {
+        currentLevel = i;
+        return;
+      }
+    });
+    if (currentLevel !== -1) { return; }
+    this.sort(property, this.data.query.sort.length, descending);
+  };
+
+  /**
+   * Remove a sort entry, shorthand of Logic.prototype.sort
+   * @param {String} property The property to remove to the sort
+   */
+  Logic.prototype.removeSort = function (property) {
+    this.sort(property, -1);
+  };
+
+
+
+  /**
+   * Get the filter descriptor associated to a property
+   * @param {String} property
+   */
+  Logic.prototype.getFilterDescriptor = function (property) {
+    if (this.data.query.properties.indexOf(property) === -1) { return; }
+    var propertyFilter = this.data.meta.propertyDescriptors.filter || {};
+    var propertyType = propertyFilter.type || "text";
+    return this.data.meta.filters[propertyType];
+  };
+
+
+  /**
+   * Get the default filter operator associated to a property
+   * @param {String} property
+   */
+  Logic.prototype.getFilterDefaultOperator = function (property) {
+    var filterDescriptor = this.getFilterDescriptor(property);
+    if (!filterDescriptor) { return; }
+    var filterOperators = filterDescriptor.operators;
+    if (!(filterOperators instanceof Array)) { return; }
+    return filterOperators[0];
+  };
+
+
+  /**
+   * Update sort configuration based on parameters, then fetch new data
+   * @param {String} property The property to filter according to
+   * @param {String} index The index of the filter entry
+   * @param {String} filterEntry The filter data used to update the filter configuration
+   *  filterEntry = {property, operator, value}
+   *  undefined values are defaulted to current values, then to default values.
+   * @param {String} filterEntry.property The new property to filter according to
+   * @param {String} filter.operator The operator of the filter.
+   *  Should match the filter descriptor of the filter property
+   * @param {String} filter.value Value for the new filter entry
+   */
+  Logic.prototype.filter = function (property, index, filterEntry) {
```

Comment:
Yes but with your use cases, you have to overwrite the whole filter object each time, you can't modify only one filter at a time. Or you can by taking the whole filter object, finding the desired filter, modify it, then pass the modified object to the filter API, but in that case it would be just simpler to implement such a system in the API.
What I can do is to add a method for filtering (and sorting too) that takes a whole object to overwrite the current configuration.

Reply:
About the `property` inside `filterEntry`, it's to support the case where the user wants to change the property that is used for the filter, as this might be useful for some layouts (but not in the table one, as the filters are inherently associated with columns).

Reply:
> You can't modify only one filter at a time.

I'm not convinced that you need this. I feel it adds unneeded complexity. The way I view it, a filter module gets as input the property name and descriptor and outputs (through an event or by calling the logic directly) the list of constraints for that property, taking into account the property descriptor. I don't find it natural to have to add each constraint individually in the filter module. I find it simpler if the filter module sets all the constraints for that property at once.

> it's to support the case where the user wants to change the property that is used for the filter

I don't visualize this. How exactly would the user change from the UI the property but keep the constraints? Another property comes with a different descriptor and probably a different filter configuration that might make some of the constraints invalid. It feels too complex.

Reply:
> I find it simpler if the filter module sets all the constraints for that property at once.

But when you are adding / modifying a filter, you're going to use an input / select / whatever plugin to specify the value you want, and you are only specifying the value for this filter only. That means that you'll have to merge this value returned by the filter element to the whole object, before calling the Logic API.
And in that case, to merge the modified value with the other filter values, I see two possible solutions:
- trying to merge somehow the value with the `data.query.filters[property]` object, but without the index of the filter we are modifying it's going to be difficult, or even impossible
- computing the filter object from scratch by getting all the filter values from every inputs / selects / plugins of the property. But in this case, the filter inputs are no longer independents and we need to create a "wrapper" module around those inputs that computes for the whole property the object to pass as an argument of the API filter method.
Also I think we are doing too much work just to compute from scratch what is going to send to the API, before the API does a similar amount of work.

> How exactly would the user change from the UI the property but keep the constraints?

For instance if you take notion filter system, you can specify first the property, then the operator and the value. If you want to modify the property, you can directly do that, and if the old and the new property filters are compatibles, it keeps the operator and the value (or just the operator). A layout could implement such a system, or even the default filter panel.
And I think it's quite easy to keep operators / values between property filters, as filter plugins are going to keep the operator and value only if it supports it, or reset it to default otherwise.

Reply:
> computing the filter object from scratch by getting all the filter values from every inputs / selects / plugins of the property

First of all, there should be a **single** filter widget per property. In other words: a single place in the live data UI where the user can filter a given property. This filter widget can enforce 0, 1 or more constraints on its target property. Look at how you can filter JIRA issues. First of all you have to choose between simple and advanced filtering. You can't use both at the same time. Then, with simple filtering you have to choose the properties to filter on from a drop down list. When you choose a property a filter widget is added to the page. You can do this only **once** per property, meaning that you can't have two filter widgets for the same property. The filter widget that was added can support single selection (single constraint) or multiple selection (multiple constraint). In the end the filter widget knows what constraints the user has selected.

> If you want to modify the property, you can directly do that, and if the old and the new property filters are compatibles, it keeps the operator and the value (or just the operator)

I still find it too complex for such a small benefit. It involves many ``IF``s to see if the property types are compatible, if the operators are compatible, if the values make sense for the new property, which means many points of failure. And I'm not convinced that the users will change the property rather than adding removing the filter. In any case I wouldn't start with that. You need to find a balance between having a design that is future proof (meaning that can be extended easily later) and writing code from the start that may never be used.

Reply:
> First of all, there should be a single filter widget per property. In other words: a single place in the live data UI where the user can filter a given property. This filter widget can enforce 0, 1 or more constraints on its target property

I don't remember we ever talked about something like this. We decided to integrate the operator inside the widget, but we never talked about putting **every** filter inside only one widget associated to the property. Also, we agreed that the user should be able to enter the value of the filter without the widget, for instance if you want to specify a date range, you can use the widget to help you, or directly write the value in the input. If this is still the case for the "complex" widget that you are talking about, then the syntax that the user will have to enter on this single input will be way more complicated. Furthermore in this case, why not just to provide the string value of this input to the filter API, instead of the complex object that you described before (like `[{value: 'mfl', operator: 'prefix'}, {value: 'Dee', operator: 'suffix'}]`) ?

> I still find it too complex for such a small benefit. It involves many IFs to see if the property types are compatible, if the operators are compatible, if the values make sense for the new property, which means many points of failure.

Well I thought that was something that would be present in every filter. If you want to filter a number but you enter some text, the filter should not allow this and not send a request to the server to fetch new data. Same for operators, if you somehow specify an operator that is not supported by the current filter, it should not allow this and fallback to a default operator. In my opinion this is not extra features, but an security to make sure the data the user provides is coherent with the logic behind. Maybe we can left that for later as you proposed, but I think this is something that should be implemented.

Reply:
> we agreed that the user should be able to enter the value of the filter without the widget, for instance if you want to specify a date range, you can use the widget to help you, or directly write the value in the input.

The text input is part of the filter widget. The fact that the date filter allows you to type the date is a feature of the date filter widget.

> If this is still the case for the "complex" widget that you are talking about, then the syntax that the user will have to enter on this single input will be way more complicated.

I don't understand this part. The user doesn't have to enter any complex syntax. That's why you have a filter widget.